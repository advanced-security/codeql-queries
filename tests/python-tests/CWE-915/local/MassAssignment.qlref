CWE-915/MassAssignmentLocal.ql