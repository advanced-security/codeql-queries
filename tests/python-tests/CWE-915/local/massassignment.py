from argparse import ArgumentParser

# Inputs
parser = ArgumentParser(__name__)
parser.add_argument("-k")
parser.add_argument("-v")

arguments = parser.parse_args()


class Users(object):
    uid: int
    username: str
    password: str

    def set(self, key: str, vaule: str):
        # codeql: py/mass-assignment
        self.__setattr__(key, vaule)


user = Users()

# codeql: py/mass-assignment
setattr(user, arguments.k, arguments.v)

# codeql: py/mass-assignment
user.__setattr__(arguments.k, arguments.v)

# Issue in the `set()` function
user.set(arguments.k, arguments.v)

# false-positive: variable isn't user controlled
setattr(user, "uid", arguments.v)
# same as: user.uid = i

# false-positive: variable isn't user controlled
user.__setattr__("uid", arguments.v)
# same as: user.uid = i
