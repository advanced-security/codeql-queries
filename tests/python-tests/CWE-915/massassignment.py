from flask import Flask, request, abort

app = Flask(__name__)


class Users(object):
    uid: int
    username: str
    password: str


@app.route("/test1")
def test1():
    # Test 1: Simple test case
    i = request.args.get("i")
    user = Users()
    
    # codeql: py/mass-assignment
    setattr(user, request.args.get("k"), request.args.get("v"))
    
    # codeql: py/mass-assignment
    user.__setattr__(request.args.get("k"), request.args.get("v"))

    # false-positive: variable isn't user controlled
    setattr(user, "uid", request.args.get("uid"))
    # same as: user.uid = i
    
    # false-positive: variable isn't user controlled
    user.__setattr__("uid", request.args.get("uid"))
    # same as: user.uid = i
    
    return f"<h1>Test 3</h1>"


@app.route("/test3")
def test3():
    user = Users()
    # Test 2:
    # remote-flow: requrst.args
    for param_name, param_vaule in request.args.items():
        # codeql: py/mass-assignment
        setattr(user, param_name, param_vaule)
    
    return f"<h1>Test 3</h1>"


@app.route("/test4")
def test4():
    # Test 3:
    if request.method == "PUT":
        if not request.json:
                abort(400)

        user_data = request.json
        user = Users()

        # remote-flow: requrst.args
        for item in user_data:
            # codeql: py/mass-assignment
            setattr(user, item, request.json[item])

    return f"<h1>Test 4</h1>"
