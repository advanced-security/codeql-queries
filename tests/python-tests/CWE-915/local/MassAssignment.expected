edges
| massassignment.py:16:19:16:21 | ControlFlowNode for key | massassignment.py:18:26:18:28 | ControlFlowNode for key |
| massassignment.py:30:10:30:20 | ControlFlowNode for Attribute | massassignment.py:16:19:16:21 | ControlFlowNode for key |
nodes
| massassignment.py:16:19:16:21 | ControlFlowNode for key | semmle.label | ControlFlowNode for key |
| massassignment.py:18:26:18:28 | ControlFlowNode for key | semmle.label | ControlFlowNode for key |
| massassignment.py:24:15:24:25 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:27:18:27:28 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:30:10:30:20 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
subpaths
#select
| massassignment.py:18:26:18:28 | ControlFlowNode for key | massassignment.py:30:10:30:20 | ControlFlowNode for Attribute | massassignment.py:18:26:18:28 | ControlFlowNode for key | Use of $@. | massassignment.py:30:10:30:20 | ControlFlowNode for Attribute | mass assignment |
| massassignment.py:24:15:24:25 | ControlFlowNode for Attribute | massassignment.py:24:15:24:25 | ControlFlowNode for Attribute | massassignment.py:24:15:24:25 | ControlFlowNode for Attribute | Use of $@. | massassignment.py:24:15:24:25 | ControlFlowNode for Attribute | mass assignment |
| massassignment.py:27:18:27:28 | ControlFlowNode for Attribute | massassignment.py:27:18:27:28 | ControlFlowNode for Attribute | massassignment.py:27:18:27:28 | ControlFlowNode for Attribute | Use of $@. | massassignment.py:27:18:27:28 | ControlFlowNode for Attribute | mass assignment |
